Implement "support doc comments for internal classes and functions"

kocsismate opened the pull request:
Fixes #13130

The biggest change of this PR (besides the internal structures of functions and classes) is how function entries are generated by gen_stub.php: so far the higher level macros were used (`ZEND_ME `, `ZEND_FALIAS` etc.), but since the variations we have has already started to skyrocket, I decided to go with the `ZEND_RAW_FENTRY ` macro which is the lowest level... As we now have a decent framework for generating the arginfo.h file by taking into account older PHP versions as well, I think this change is less problematic... A random example:

```
#if (PHP_VERSION_ID >= 80400)
	ZEND_RAW_FENTRY("zend_test_zend_ini_parse_quantity", zif_zend_test_zend_ini_parse_quantity, arginfo_zend_test_zend_ini_parse_quantity, 0, NULL)
#else
	ZEND_RAW_FENTRY("zend_test_zend_ini_parse_quantity", zif_zend_test_zend_ini_parse_quantity, arginfo_zend_test_zend_ini_parse_quantity, 0)
#endif
```

(`NULL`  stands for the `doc_comment` in the function entry generated for PHP 8.4+).

kocsismate commented on .gitattributes:
```diff
@@ -20,7 +20,7 @@
 
 # Collapse generated files within git and pull request diff.
-**/*_arginfo.h linguist-generated -diff
+#**/*_arginfo.h linguist-generated -diff
```
this is just a temporary change in order to make the review easier

kocsismate commented:
It looks like I managed to fix all the test failures so I'd appreciate any reviews. :)

ju1ius commented:
@kocsismate Works for me!

I would prefer a solution not involving `zend_function_entry`, but IDK if that would be more complex to implement on the `gen_stubs` side, so...

👍🏻

kocsismate commented:
> However, we may at least try to retain the most simple ones, like ZEND_FE/ZEND_ME when there are no special modifiers. Otherwise, it may be easier to just use struct initializers directly and omit the default fields.

Hm, that doesn't sound a bad idea. :) I'm a little bit hesitant because doing so will complicate gen_stub.php adding yet another `if`s. But I'll try to implement it sometime soon.

kocsismate commented on Zend/zend_compile.h:
```diff
@@ -524,6 +524,7 @@ typedef struct _zend_internal_function {
 	/* END of common elements */
 
 	zif_handler handler;
+	zend_string *doc_comment;
```
good idea! I missed this opportunity

kocsismate commented on Zend/zend_opcode.c:
```diff
@@ -419,6 +419,10 @@ ZEND_API void destroy_zend_class(zval *zv)
 		case ZEND_INTERNAL_CLASS:
+			if (ce->doc_comment) {
+				zend_string_release_ex(ce->doc_comment, 1);
+			}
```
Yes, I didn't find the function destructor first and then forgot about it. I hope it's ok now

kocsismate commented:
@iluuu1994 I slightly refactored the behavior, and now `ZEND_FE` and `ZEND_ME` is also generated if possible. This also means that I got rid of generating `ZEND_FRAMELESS_FE`s because this use-case also deemed to be uncommon.

When implementing this, I added support for conditional frameless func info item generation based on the PHP version compatibility requirements. I didn't commit this fix separately because it needed the PHP version compatibility info which I had just added in this PR.

TimWolla commented on ext/zend_test/test.c:
```diff
@@ -761,7 +763,8 @@ static zend_function *zend_test_class_method_get(zend_object **object, zend_stri
 	    fptr->fn_flags = ZEND_ACC_CALL_VIA_HANDLER;
 	    fptr->function_name = zend_string_copy(name);
-	    fptr->handler = ZEND_FN(zend_test_func);
+		fptr->handler = ZEND_FN(zend_test_func);
+		fptr->doc_comment = NULL;
```
Inconsistent indentation here.

devnexen commented:
pgsql looks ok to me.

kocsismate commented:
Ilija, could you have another look please?

iluuu1994 commented on ext/reflection/php_reflection.c:
```diff
@@ -889,6 +895,9 @@ static zval *property_get_default(zend_property_info *prop_info) {
 /* {{{ _property_string */
 static void _property_string(smart_str *str, zend_property_info *prop, const char *prop_name, char* indent)
 {
+	if (prop && prop->doc_comment) {
```
Same here.

iluuu1994 reviewed (approved):
The C code LGTM! I didn't look closely at `build/gen_stub.php`.

Girgias reviewed (approved):
LGTM, also didn't check gen_stub too closely

kocsismate commented on ext/reflection/php_reflection.c:
```diff
@@ -566,6 +566,10 @@ static void _class_const_string(smart_str *str, zend_string *name, zend_class_co
 	const char *type = type_str ? ZSTR_VAL(type_str) : zend_zval_type_name(&c->value);
+
+	if (c->doc_comment) {
+		smart_str_append_printf(str, "%s%s\n", indent, ZSTR_VAL(c->doc_comment));
```
Ah yes, these are unrelated changes indeed. I completely forgot that I did these. I can move them, sure.

kocsismate replied:
Filed and merged https://github.com/php/php-src/pull/13499
